independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/project/QuarkusProjectHelper.java

```diff
@@ -91,47 +102,60 @@ public static ExtensionCatalog resolveExtensionCatalog() throws RegistryResoluti
+    }
+
+    public static List<ResourceLoader> getCodestartResourceLoaders(ExtensionCatalog catalog,
+            MavenArtifactResolver mvn) {
+        return getCodestartResourceLoaders(BASE_CODESTART_ARTIFACT_COORDS, catalog, mvn);
+    }
+
+    public static List<ResourceLoader> getCodestartResourceLoaders(String baseCodestartArtifactCoords, ExtensionCatalog catalog,
+            MavenArtifactResolver mvn) {
+        final Map<String, Artifact> codestartsArtifacts = new LinkedHashMap<>();
+        if (catalog != null) {
+            // Load codestarts from each extensions codestart artifacts
+            for (Extension e : catalog.getExtensions()) {
+                final String artifactCoords = ExtensionProcessor.of(e).getCodestartArtifact();
+                if (artifactCoords == null || codestartsArtifacts.containsKey(artifactCoords)) {
+                    continue;
+                }
+                codestartsArtifacts.put(artifactCoords, DependencyNodeUtils.toArtifact(artifactCoords));
             }
-            codestartArtifacts.put(artifactCoords, DependencyNodeUtils.toArtifact(artifactCoords));
-        }
-        Object o = catalog.getMetadata().get(CODESTARTS_ARTIFACTS);
-        if (o != null) {
-            @SuppressWarnings({ "unchecked" })
-            final List<Object> list = o instanceof List ? (List<Object>) o : Arrays.asList(o);
-            for (Object i : list) {
-                final String artifactCoords = i.toString();
-                if (codestartArtifacts.containsKey(artifactCoords)) {
+
+            // Load codestarts from catalog codestart artifacts
+            final List<String> catalogCodestartArtifacts = getCodestartArtifacts(catalog);
+            for (String artifactCoords : catalogCodestartArtifacts) {
+                if (codestartsArtifacts.containsKey(artifactCoords)) {
                     continue;
                 }
-                codestartArtifacts.put(artifactCoords, DependencyNodeUtils.toArtifact(artifactCoords));
+                codestartsArtifacts.put(artifactCoords, DependencyNodeUtils.toArtifact(artifactCoords));
             }
         }
+        // Load codestarts from the base artifact
+        codestartsArtifacts.put(baseCodestartArtifactCoords, DependencyNodeUtils.toArtifact(baseCodestartArtifactCoords));
 
-        final URL[] urls = new URL[codestartArtifacts.size()];
-        int i = 0;
-        for (Artifact a : codestartArtifacts.values()) {
+        final List<ResourceLoader> codestartResourceLoaders = new ArrayList<>(codestartsArtifacts.size());
+        for (Artifact a : codestartsArtifacts.values()) {
             try {
-                urls[i++] = mvn.resolve(a).getArtifact().getFile().toURI().toURL();
+                final URL artifactUrl = mvn.resolve(a).getArtifact().getFile().toURI().toURL();
+                final ClassPathResourceLoader resourceLoader = new ClassPathResourceLoader(
+                        new URLClassLoader(new URL[] { artifactUrl }, null));
```

Comment:
I think we could avoid classloaders here. These artifacts resolve to either a JAR or directory. Avoiding URLs will help supporting native image.

Reply:
Yes, exactly, that's the big problem with the current code. If things are in external jars, we need to load them from the files, not using a CL.

Reply:
@gsmet @aloubyansky  I didn't change that behavior, it was already using the CL and URL. I think this PR is big enough and we should provide another PR for this?

Reply:
Sure

Reply:
Fixed because it may be the cause of CI failing on Windows (URLClassLoader is closeable)